lib/Dialect/Seq/SeqOps.cpp

```diff
@@ -219,6 +219,81 @@ void HLMemOp::build(OpBuilder &builder, OperationState &result, Value clk,
   HLMemOp::build(builder, result, t, clk, rst, symName);
 }
 
+//===----------------------------------------------------------------------===//
+// FIFOOp
+//===----------------------------------------------------------------------===//
+
+// Flag threshold custom directive
+static ParseResult parseFIFOFlagThreshold(OpAsmParser &parser,
+                                          IntegerAttr &threshold,
+                                          Type &outputFlagType,
+                                          StringRef directive) {
+  // look for an optional "almost_full $threshold>" group.
+  if (succeeded(parser.parseOptionalKeyword(directive))) {
+    int64_t thresholdValue;
+    if (succeeded(parser.parseInteger(thresholdValue))) {
+      threshold = parser.getBuilder().getI64IntegerAttr(thresholdValue);
+      outputFlagType = parser.getBuilder().getI1Type();
+      return success();
+    }
+    return parser.emitError(parser.getNameLoc(),
+                            "expected integer value after " + directive +
+                                " directive");
+  }
+  return success();
+}
+
+ParseResult parseFIFOAFThreshold(OpAsmParser &parser, IntegerAttr &threshold,
+                                 Type &outputFlagType) {
+  return parseFIFOFlagThreshold(parser, threshold, outputFlagType,
+                                "almost_full");
+}
+
+ParseResult parseFIFOAEThreshold(OpAsmParser &parser, IntegerAttr &threshold,
+                                 Type &outputFlagType) {
+  return parseFIFOFlagThreshold(parser, threshold, outputFlagType,
+                                "almost_empty");
+}
+
+static void printFIFOFlagThreshold(OpAsmPrinter &p, Operation *op,
```

Comment:
As a rule, if the amount of extra code is more than the code you're sharing, don't share it.